// @vavt/util/lib/es/index.mjs
const A = (e) => {
  const n = typeof e;
  return n !== "function" && n !== "object" || e === null;
}, E = (e) => {
  const n = e.flags === "" ? void 0 : e.flags;
  return new RegExp(e.source, n);
}, a = (e, n = /* @__PURE__ */ new WeakMap()) => {
  if (e === null || A(e))
    return e;
  if (n.has(e))
    return n.get(e);
  if (e instanceof RegExp)
    return E(e);
  if (e instanceof Date)
    return new Date(e.getTime());
  if (e instanceof Function)
    return e;
  if (e instanceof Map) {
    const o = /* @__PURE__ */ new Map();
    return n.set(e, o), e.forEach((r, c) => {
      o.set(c, a(r, n));
    }), o;
  }
  if (e instanceof Set) {
    const o = /* @__PURE__ */ new Set();
    n.set(e, o);
    for (const r of e)
      o.add(a(r, n));
    return o;
  }
  if (Array.isArray(e)) {
    const o = [];
    return n.set(e, o), e.forEach((r) => {
      o.push(a(r, n));
    }), o;
  }
  const t = {};
  n.set(e, t);
  for (const o in e)
    Object.prototype.hasOwnProperty.call(e, o) && (t[o] = a(e[o], n));
  return t;
}, M = (e, n = 200) => {
  let t = 0;
  return (...o) => new Promise((r) => {
    t && (clearTimeout(t), r("cancel")), t = window.setTimeout(() => {
      e.apply(void 0, o), t = 0, r("done");
    }, n);
  });
}, x = (e, n = {
  _blank: !0,
  nofollow: !0
}) => {
  const t = document.createElement("a");
  t.href = e, n._blank && (t.target = "_blank"), n.nofollow && (t.rel = "noopener noreferrer"), t.click();
}, C = (e, n = "") => {
  const t = document.createElement("a");
  t.setAttribute("href", e), t.setAttribute("download", n), t.click();
}, T = () => {
  let e = -1;
  return (t, o, r, c = 100) => {
    const s = () => {
      r && (typeof c == "number" ? setTimeout(r, c) : r());
    };
    e !== -1 && (cancelAnimationFrame(e), s());
    let i = t.scrollTop;
    const l = () => {
      e = -1;
      const u = o - i;
      i = i + u / 5, Math.abs(u) < 1 ? (t.scrollTo(0, o), s()) : (t.scrollTo(0, i), e = requestAnimationFrame(l));
    };
    e = requestAnimationFrame(l);
  };
}, R = T(), L = (e, n = 200) => {
  let t = 0, o = null;
  return (...r) => {
    const c = (s) => {
      t === 0 && (t = s), s - t >= n ? (e.apply(void 0, o), o = null, t = 0) : window.requestAnimationFrame(c);
    };
    o === null && window.requestAnimationFrame(c), o = r;
  };
}, F = (e, n = !0) => {
  if (!e)
    return {};
  const t = {};
  return e.replace(/^\?/, "").split("&").forEach((o) => {
    const [r, c] = o.split("=");
    if (r) {
      const s = n ? decodeURIComponent(r) : r, i = n ? decodeURIComponent(c) : c;
      t[s] !== void 0 ? (Array.isArray(t[s]) || (t[s] = [t[s]]), t[s].push(i)) : t[s] = i;
    }
  }), t;
}, H = (e, n = !0) => e ? Object.keys(e).map((t) => {
  const o = n ? encodeURIComponent(t) : t;
  if (Array.isArray(e[t]))
    return e[t].map((r) => {
      const c = n ? encodeURIComponent(r) : r;
      return `${o}=${c}`;
    }).join("&");
  {
    const r = n ? encodeURIComponent(e[t]) : e[t];
    return `${o}=${r}`;
  }
}).join("&") : "", U = (e, n, t = "asc") => e.sort((o, r) => {
  const c = n(o), s = n(r);
  return c === void 0 && s !== void 0 || c === null && s !== void 0 ? t === "asc" ? -1 : 1 : c == null && s === void 0 || s == null ? t === "asc" ? 1 : -1 : t == "asc" ? String(c).localeCompare(String(s), void 0, {
    numeric: !0
  }) : String(s).localeCompare(String(c), void 0, {
    numeric: !0
  });
}), _ = (e) => {
  const n = (t) => {
    const { scrollHeight: o, scrollWidth: r, offsetHeight: c, offsetWidth: s, scrollLeft: i, scrollTop: l } = e, u = t.x, w = t.y, p = (g) => {
      const f = l + w - g.y, m = i + u - g.x, S = o - c, y = r - s, d = {};
      m >= 0 && m <= y && (d.left = m), f >= 0 && f <= S && (d.top = f), e.scroll(d);
    };
    document.addEventListener("mousemove", p);
    const v = () => {
      document.removeEventListener("mousemove", p), document.removeEventListener("mouseup", v);
    };
    document.addEventListener("mouseup", v);
  };
  return e.addEventListener("mousedown", n), () => {
    e.removeEventListener("mousedown", n);
  };
}, $ = () => `${Date.now().toString(36)}${Math.random().toString(36).substring(2)}`, k = (e) => !isNaN(Number(e));
export {
  T as createSmoothScroll,
  M as debounce,
  a as deepClone,
  C as download,
  _ as draggingScroll,
  k as isNumber,
  A as isPrimitive,
  x as linkTo,
  H as objToSearch,
  U as objectSort,
  F as searchToObj,
  R as smoothScroll,
  L as throttle,
  $ as uuid
};
